tests/Tests.Shared/test-api.h

```diff
@@ -0,0 +1,583 @@
+    static std::mutex s_alternateTestNameLock;
+    std::lock_guard<std::mutex> lock(s_alternateTestNameLock);
+    std::string toReturn = s_alternateTestName;
+    if (exchange) {
+        s_alternateTestName = alternateName;
+    }
+    
+    return toReturn;
+}
+
+#pragma push_macro("GTEST_TEST_")
+#define GTEST_TEST_(className, methodName, parent_class, parent_id) \
+    class className##_##methodName : public parent_class { \
+        BEGIN_TEST_CLASS(className##_##methodName) \
+            TEST_CLASS_PROPERTY(L"Ignore", _IsDisabledTest( #methodName )) \
+        END_TEST_CLASS() \
+    \
+        TEST_METHOD(Test); \
+        TEST_METHOD_SETUP(_Setup); \
+        TEST_METHOD_CLEANUP(_Teardown); \
+        \
+        \
+        virtual void TestBody() { \
+            Test(); \
+        } \
+        \
+    }; \
+    \
+    bool className##_##methodName::_Setup() { \
+        SetUp(); \
+        return true; \
+    } \
+    bool className##_##methodName::_Teardown() { \
+        TearDown(); \
+        return true; \
+    } \
+    void className##_##methodName::Test()
+
+#pragma push_macro("INSTANTIATE_TEST_CASE_P")
+#define  INSTANTIATE_TEST_CASE_P(prefix, className, generator, ...)                  \
+/* This assumes that there already exists a className type with SetUp and TearDown methods to call */ \
+    \
+    class prefix##_##className { \
+        BEGIN_TEST_CLASS(prefix##_##className) \
+                TEST_CLASS_PROPERTY(L"Ignore", _IsDisabledTest( #prefix )) \
+        END_TEST_CLASS() \
+    \
+        TEST_METHOD(Test) { \
+            WEX::Logging::Log::Comment(L"\nStarting parameterized subtests for " L#prefix L"_" L#className L":"); \
+            ::testing::internal::ParamGenerator<className::ParamType> paramGenerator(generator); \
+            for (const auto& metaFactory : ::testing::UnitTest::GetInstance()->parameterized_test_registry(). \
+              GetTestCasePatternHolder<className>( \
+                  #className, \
+                  ::testing::internal::CodeLocation( \
+                      __FILE__, __LINE__))->GetTestMetaFactories()) { \
+                 size_t i = 0; \
+                 static std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter; \
+                 static const std::wstring c_stringPrefix = L #prefix L"_" L#className L"." + converter.from_bytes(metaFactory.first.c_str()); \
+                 for (typename ::testing::internal::ParamGenerator<className::ParamType>::iterator param = paramGenerator.begin(); param != paramGenerator.end(); ++param, ++i) { \
+                    std::wstring currentTest = c_stringPrefix + L"/" + std::to_wstring(i); \
+                    /* TODO: don't leak. Use better storage. */ \
+                    auto factory = metaFactory.second->CreateTestFactory(*param); \
+                    auto test = factory->CreateTest(); \
+                    \
+                    WEX::Logging::Log::StartGroup(currentTest.c_str()); \
```

Comment:
>WEX::Logging::Log::StartGroup(currentTest.c_str()); \ [](start = 20, length = 53)

would be useful to also log what the parameters used here are. I think gtest has a method for that somewhere. I dunno if I can get to it off of the metafactory though?